src/misc/coloring.c

```diff
@@ -129,6 +133,135 @@ static igraph_error_t igraph_i_vertex_coloring_greedy_cn(const igraph_t *graph,
     return IGRAPH_SUCCESS;
 }
 
+/* DSatur coloring heuristics */
+
+typedef struct {
+    igraph_integer_t saturation_degree;
+    igraph_integer_t edge_degree;
+} igraph_color_metadata_t;
+
+static int igraph_i_color_metadata_compate(const void* left, const void* right){
+    igraph_color_metadata_t left_d  = *((igraph_color_metadata_t*)left);
+    igraph_color_metadata_t right_d = *((igraph_color_metadata_t*)right);
+    if(left_d.saturation_degree==right_d.saturation_degree){
+        return left_d.edge_degree - right_d.edge_degree;
+    }
+    return left_d.saturation_degree - right_d.saturation_degree;
+}
+
+
+
+static igraph_bool_t is_color_used_by_neighbour(
+    const igraph_vector_int_t *colors, igraph_integer_t color,
+    const igraph_vector_int_t *neighbours
+) {
+    igraph_integer_t nbr_cnt = igraph_vector_int_size(neighbours);
+
+    for (igraph_integer_t nbr_indx = 0; nbr_indx < nbr_cnt; nbr_indx++) {
+        igraph_integer_t nbr = VECTOR(*neighbours)[nbr_indx];
+        if (VECTOR(*colors)[nbr] == color) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+static void igraph_i_dsatur_update_heap(
+    const igraph_adjlist_t *adjlist, igraph_gen2wheap_t *node_degrees_heap, 
+    igraph_vector_int_t *neighbours, igraph_vector_int_t *colors, 
+    igraph_integer_t color
+) {
+    igraph_gen2wheap_delete_max(node_degrees_heap);
+    igraph_color_metadata_t color_metadata; 
+    igraph_integer_t nbr_cnt = igraph_vector_int_size(neighbours);
+    for (igraph_integer_t nbr_indx = 0; nbr_indx < nbr_cnt; nbr_indx++){
+        igraph_integer_t nbr = VECTOR(*neighbours)[nbr_indx];
+        color_metadata = *((igraph_color_metadata_t*)igraph_gen2wheap_get(node_degrees_heap, nbr));
```

Comment:
This crashes because you are trying to access an element without checking whether it is actually contained within the heap. I suggest you take a look at how the `IGRAPH_COLORING_GREEDY_COLORED_NEIGHBORS` is implemented and mirror that. Most of the greedy colouring approaches, _including DSatur_, will be very similar to each other and can be implemented in similar ways.